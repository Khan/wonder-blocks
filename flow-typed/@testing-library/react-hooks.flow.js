// @flow
/**
 * This file is based on:
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/types/index.ts
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/types/react.ts
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/pure.ts
 */
 declare module "@testing-library/react-hooks" {
    declare export type Renderer<TProps> = {|
        render: (props?: TProps) => void,
        rerender: (props?: TProps) => void,
        unmount: () => void,
        act: Act,
    |};

    declare export type ServerRenderer<TProps> = {|
        ...Renderer<TProps>,
        hydrate: () => void,
    |};

    declare export type RendererProps<TProps, TResult> = {|
        callback: (props: TProps) => TResult,
        setError: (error: Error) => void,
        setValue: (value: TResult) => void,
    |};

    declare export type CreateRenderer<
        TProps,
        TResult,
        TRendererOptions: {...},
        TRenderer: Renderer<TProps>,
    > = (
        props: RendererProps<TProps, TResult>,
        options: TRendererOptions,
    ) => TRenderer;

    declare export type RenderResult<TValue> = {|
        +all: Array<TValue | Error>,
        +current: TValue,
        +error?: Error,
    |};

    declare export type ResultContainer<TValue> = {|
        result: RenderResult<TValue>,
    |};

    declare export type WaitOptions = {|
        interval?: number | false,
        timeout?: number | false,
    |};

    declare export type WaitForOptions = WaitOptions;
    declare export type WaitForValueToChangeOptions = WaitOptions;
    declare export type WaitForNextUpdateOptions = WaitOptions["timeout"];

    declare export type WaitFor = (
        callback: () => boolean | void,
        options?: WaitForOptions,
    ) => Promise<void>;
    declare export type WaitForValueToChange = (
        selector: () => mixed,
        options?: WaitForValueToChangeOptions,
    ) => Promise<void>;
    declare export type WaitForNextUpdate = (
        options?: WaitForNextUpdateOptions,
    ) => Promise<void>;

    declare export type AsyncUtils = {|
        waitFor: WaitFor,
        waitForValueToChange: WaitForValueToChange,
        waitForNextUpdate: WaitForNextUpdate,
    |};

    declare export type RenderHookResult<
        TProps,
        TValue,
        TRenderer: Renderer<TProps> = Renderer<TProps>,
    > = {|
        ...ResultContainer<TValue>,
        // Omit<Renderer<TProps>, 'render' | 'act'>
        ...$Diff<
            Renderer<TProps>,
            {|
                render: Renderer<TProps>["render"],
                act: Renderer<TProps>["act"],
            |},
        >,
        // Omit<TRenderer, keyof Renderer<TProps>>
        ...$Diff<TRenderer, Renderer<TProps>>,
        ...AsyncUtils,
    |};

    declare export type ServerRenderHookResult<
        TProps,
        TValue,
        TRenderer: ServerRenderer<TProps> = ServerRenderer<TProps>,
    > = RenderHookResult<TProps, TValue, TRenderer>;

    declare export type Act = {|
        (callback: () => Promise<void>): Promise<void>,
        (callback: () => void): void,
    |};

    declare export type CleanupCallback = () => Promise<void> | void;

    declare export type WrapperComponent<TProps> = React$ComponentType<TProps>;

    declare export type RendererOptions<TProps> = {|
        wrapper?: WrapperComponent<TProps>,
    |};

    declare export type RenderHookOptions<TProps> = {|
        initialProps?: TProps,
        wrapper?: WrapperComponent<any>,
    |};

    declare export type ReactHooksRenderer = {|
        renderHook: <TProps, TResult>(
            callback: (props: TProps) => TResult,
            options?: RenderHookOptions<TProps>,
        ) => RenderHookResult<TProps, TResult>,
        act: Act,
        cleanup: () => Promise<void>,
        addCleanup: (callback: CleanupCallback) => () => void,
        removeCleanup: (callback: CleanupCallback) => void,
        suppressErrorOutput: () => () => void,
    |};

    declare export type ReactHooksServerRenderer = {|
        // Omit<ReactHooksRenderer, 'renderHook'>
        ...$Diff<
            ReactHooksRenderer,
            {|
                renderHook: ReactHooksRenderer["renderHook"],
            |},
        >,
        renderHook: <TProps, TResult>(
            callback: (props: TProps) => TResult,
            options?: RenderHookOptions<TProps>,
        ) => ServerRenderHookResult<TProps, TResult>,
    |};

    declare export function renderHook<TProps, TResult>(
        callback: (props: TProps) => TResult,
        options?: RenderHookOptions<TProps>,
    ): RenderHookResult<TProps, TResult>;
    declare export var act: Act;
    declare export function cleanup(): Promise<void>;
    declare export function addCleanup(callback: CleanupCallback): () => void;
    declare export function removeCleanup(callback: CleanupCallback): void;
    declare export function suppressErrorOutput(): () => void;
}
