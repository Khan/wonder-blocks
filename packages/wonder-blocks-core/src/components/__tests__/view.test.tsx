/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
import * as React from "react";
import {render} from "@testing-library/react";

import View from "../view";

describe("View", () => {
    it("Should set the tag to be section", () => {
        // Arrage, Act
        const view = <View tag="section" />;
        const {container} = render(view);

        const section = container.querySelector("section");

        // Assert
        expect(container).toMatchSnapshot();
        expect(section).toBeDefined();
    });

    it("Should set the tag to be article", () => {
        // Arrage, Act
        const view = <View tag="article" />;
        const {container} = render(view);

        const article = container.querySelector("article");

        // Assert
        expect(container).toMatchSnapshot();
        expect(article).toBeDefined();
    });

    it("Should set the tag to be aside", () => {
        // Arrage, Act
        const view = <View tag="aside" />;
        const {container} = render(view);

        const aside = container.querySelector("aside");

        // Assert
        expect(container).toMatchSnapshot();
        expect(aside).toBeDefined();
    });

    it("Should set the tag to be nav", () => {
        // Arrage, Act
        const view = <View tag="nav" />;
        const {container} = render(view);

        const nav = container.querySelector("nav");

        // Assert
        expect(container).toMatchSnapshot();
        expect(nav).toBeDefined();
    });

    it("Should set the tag to be main", () => {
        // Arrage, Act
        const view = <View tag="main" />;
        const {container} = render(view);

        const main = container.querySelector("main");

        // Assert
        expect(container).toMatchSnapshot();
        expect(main).toBeDefined();
    });

    it("Should set the tag to be div", () => {
        // Arrage, Act
        const view = <View />;
        const {container} = render(view);

        const div = container.querySelector("div");

        // Assert
        expect(container).toMatchSnapshot();
        expect(div).toBeDefined();
    });
});
