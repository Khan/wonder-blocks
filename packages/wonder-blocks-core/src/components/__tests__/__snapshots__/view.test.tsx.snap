// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Should set the tag to be article 1`] = `
<div>
  <article
    class=""
    style="align-items: stretch; border-width: 0px; border-style: solid; box-sizing: border-box; display: flex; flex-direction: column; margin: 0px; padding: 0px; position: relative; z-index: 0; min-height: 0; min-width: 0;"
  />
</div>
`;

exports[`View Should set the tag to be aside 1`] = `
<div>
  <aside
    class=""
    style="align-items: stretch; border-width: 0px; border-style: solid; box-sizing: border-box; display: flex; flex-direction: column; margin: 0px; padding: 0px; position: relative; z-index: 0; min-height: 0; min-width: 0;"
  />
</div>
`;

exports[`View Should set the tag to be div 1`] = `
<div>
  <div
    class=""
    style="align-items: stretch; border-width: 0px; border-style: solid; box-sizing: border-box; display: flex; flex-direction: column; margin: 0px; padding: 0px; position: relative; z-index: 0; min-height: 0; min-width: 0;"
  />
</div>
`;

exports[`View Should set the tag to be main 1`] = `
<div>
  <main
    class=""
    style="align-items: stretch; border-width: 0px; border-style: solid; box-sizing: border-box; display: flex; flex-direction: column; margin: 0px; padding: 0px; position: relative; z-index: 0; min-height: 0; min-width: 0;"
  />
</div>
`;

exports[`View Should set the tag to be nav 1`] = `
<div>
  <nav
    class=""
    style="align-items: stretch; border-width: 0px; border-style: solid; box-sizing: border-box; display: flex; flex-direction: column; margin: 0px; padding: 0px; position: relative; z-index: 0; min-height: 0; min-width: 0;"
  />
</div>
`;

exports[`View Should set the tag to be section 1`] = `
<div>
  <section
    class=""
    style="align-items: stretch; border-width: 0px; border-style: solid; box-sizing: border-box; display: flex; flex-direction: column; margin: 0px; padding: 0px; position: relative; z-index: 0; min-height: 0; min-width: 0;"
  />
</div>
`;
